### PR TITLE
Adding scikit learn linear regression workflow api

### PR DESCRIPTION
**Pull request by rajithkrishnegowda**
This notebook demonstrates a scikit linear regression tutorial using the Workflow Interface

Screenshot 2024-11-21 at 4 18 49 PM

Screenshot 2024-11-21 at 4 18 59 PM

### CODE REVIEW

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "9a61e30e-83d1-422d-aa0a-ebd4376197ec",
+   "metadata": {},
+   "source": [
+    "# Scikit-learn Linear Regression Tutorial using Workflow Interface with Ridge Regularization\n",
+    "\n",
+    "\n",
+    "This tutorial demonstrates how to train a linear regression model using scikit-learn with Ridge regularization on a dataset, leveraging the new FedAI Workflow Interface. The Workflow Interface provides a novel way to compose federated learning experiments with OpenFL, enabling researchers to handle non-IID data and perform federated averaging with optimizations like FedProx. Through this tutorial, you will learn how to set up the federated learning environment, define the flow, and execute the training process across multiple collaborators\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: there is no FedProx optimizer used here, please remove it

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+   "source": [
+    "## Now we run the same training on federated learning workflow api"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "66cb4ecf-cb27-4367-891a-6cea2f1d347f",
+   "metadata": {},
+   "source": [
+    "## Test on a Federation"
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: this is not required.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65e2b0a2-5a79-4f92-b255-4c8d3e28b635",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def inference(model, test_loader, batch_size):\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: remove this method, we are not using this.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 1 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: changes done in 26834c40e44d73315398addd340c2138cf4159b7

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 2 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: changes done in 26834c40e44d73315398addd340c2138cf4159b7

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 3 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: changes done in 26834c40e44d73315398addd340c2138cf4159b7

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "from typing import List, Union\n",
+    "\n",
+    "from sklearn.linear_model import Lasso\n",
+    "from sklearn.preprocessing import StandardScaler\n",
+    "from sklearn.datasets import make_regression\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: not being used

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Import ncessary libraries\n",
+    "import numpy as np\n",
+    "from sklearn.linear_model import Lasso\n",
+    "from sklearn.metrics import mean_squared_error\n",
+    "from sklearn.datasets import make_regression\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: not being used

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+    "from openfl.experimental.interface import FLSpec\n",
+    "from openfl.experimental.placement import aggregator, collaborator\n",
+    "from openfl.experimental.runtime import FederatedRuntime\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "\n",
+    "from tensorflow.keras.utils import to_categorical\n",
```
Comment (rahulga1) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: remove to_categorical and mnist

File: openfl-tutorials/experimental/workflow_interface_requirements.txt
Comment (theakshaypant) [inline openfl-tutorials/experimental/workflow_interface_requirements.txt:1]: Are these new requirements specific to the added experiment or for all experiments?
If these are not for all the experiments, we should add them in the experiment notebook in my opinion.

File: openfl-tutorials/experimental/workflow_interface_requirements.txt [hunk 7]
```diff
@@ -2,8 +2,15 @@ astor==0.8.1
 chardet
 charset-normalizer
 dill==0.3.6
+matplotlib>=2.0.0
 metaflow==2.7.15
+mistune>=2.0.3 # not directly required, pinned by Snyk to avoid a vulnerability
 nbdev==2.3.12
+numpy>=1.13.3
+openfl>=1.2.1
```
Comment (theakshaypant) [inline openfl-tutorials/experimental/workflow_interface_requirements.txt]: Any reason for having the minimum required version for openfl to be this old?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 4 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e1e711efeb4f84ed325d0ae3401328c80502479a

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 6 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e1e711efeb4f84ed325d0ae3401328c80502479a

File: openfl-tutorials/experimental/workflow_interface_requirements.txt (hunk 7 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/workflow_interface_requirements.txt]: resolved in e1e711efeb4f84ed325d0ae3401328c80502479a

File: openfl-tutorials/experimental/workflow_interface_requirements.txt
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/workflow_interface_requirements.txt:1]: resolved in e1e711efeb4f84ed325d0ae3401328c80502479a

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 5 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e1e711efeb4f84ed325d0ae3401328c80502479a

**Review by rahulga1 — APPROVED** [review]

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+    "print(\"Test matrix shape\", X_test.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "052a4195-a410-4983-84b9-942159d9d345",
+   "metadata": {},
+   "source": [
+    "## Shard Splitter Class\n",
+    "Define a helper class to split the data into shards for federated learning."
```
Comment (tanwarsh) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: This comment does not belong to the code block below. Move it below this code block. or change it as this is not a class. It is using ShardSplitter class to split the dara

Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b3c0039-e1f7-4047-b98b-a2d4bd42f015",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Import ncessary libraries\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:238]: nit: Already these libraries are imported in the "Getting Started" section. We can remove this duplicate imports.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+    "federated_flow = FederatedLassoFlow(model=lr_model, num_rounds=10)\n",
+    "\n",
+    "# Set the runtime for federated learning\n",
+    "federated_flow.runtime = local_runtime\n",
+    "\n",
+    "# Start the federated learning process\n",
+    "federated_flow.run()"
+   ]
+  },
+  {
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:497]: Can we add one more step where final trained model is validated against random data set ?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+   "source": [
+    "# Define the federated learning workflow\n",
+    "from openfl.experimental.placement import aggregator, collaborator\n",
+    "\n",
+    "# Federated Learning Workflow using OpenFL's Workflow API\n",
+    "class FederatedLassoFlow(FLSpec):\n",
+    "    def __init__(self, model, num_rounds=3):\n",
+    "        self.model = model\n",
+    "        self.num_rounds = num_rounds\n",
+    "        self.current_round = 1\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: nit: Initialized current_round in start step, is this required here ?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+    "        mse = self.model.mse(x_test, y_test)\n",
+    "        print(f\"aggregation model validation MSE: {mse:.4f}\")\n",
+    "        self.aggregated_mse = mse\n",
+    "        self.next(self.train)\n",
+    "\n",
+    "    @collaborator\n",
+    "    def train(self):\n",
+    "        x_train, y_train = self.train_loader\n",
+    "        print(f'x_train shape: {x_train.shape}, y_train shape: {y_train.shape}')\n",
+    "        history = self.model.fit(\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: nit: This variable is not used anywhere ?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65e2b0a2-5a79-4f92-b255-4c8d3e28b635",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Federated Averaging for Lasso models\n",
+    "def FedAvg(models):\n",
+    "    new_model = models[0]\n",
+    "    coef_list = [model.model.coef_ for model in models]\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:373]: nit: Can we name the model variable something different like input_model, model.model looks odd ?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+    "class ShardSplitter:\n",
+    "    def __init__(self, num_shards):\n",
+    "        self.num_shards = num_shards\n",
+    "\n",
+    "    def split(self, X, y):\n",
+    "        \"\"\"Split the given 2D numpy arrays X and y into equal shards and return list of indexes for each shard.\"\"\"\n",
+    "        num_samples = X.shape[0]\n",
+    "        shard_size = num_samples // self.num_shards\n",
+    "        indexes = np.arange(num_samples)\n",
+    "        #np.random.shuffle(indexes)\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: nit: remove this commented line if not required.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "class SklearnLinearRegressionLasso:\n",
+    "    def __init__(self, n_feat: int, alpha: float = 1.0) -> None:\n",
+    "        self.model = Lasso(alpha=alpha)\n",
+    "        self.scaler = StandardScaler()\n",
+    "        self.weights = np.ones((n_feat + 1)) \n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: Is self.weights required ?

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 16]
```diff
@@ -0,0 +1,546 @@
...
+    "    def mse(self, X: np.ndarray, Y: np.ndarray) -> float:\n",
+    "        Y_predict = self.predict(X)\n",
+    "        return mean_squared_error(Y, Y_predict)\n",
+    "    \n",
+    "    def fit(self, X: np.ndarray, Y: np.ndarray, silent: bool=False) -> None:\n",
+    "   \n",
+    "        X = self.scaler.fit_transform(X)\n",
+    "        self.model.fit(X, Y)\n",
+    "        mse = self.mse(X, Y)\n",
+    "        #self.weights[:-1] = self.model.coef_\n",
```
Comment (kagrawa2) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: we can remove the commented code if not required.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 13 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:373]: model.coef_ variable is not user defined variable.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 9 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:238]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 10 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb:497]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 17]
```diff
@@ -0,0 +1,546 @@
...
+   "execution_count": null,
+   "id": "1b3c0039-e1f7-4047-b98b-a2d4bd42f015",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Import ncessary libraries\n",
+    "import numpy as np\n",
+    "from sklearn.linear_model import Lasso\n",
```
Comment (tanwarsh) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: duplicate imports

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 12 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 11 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 14 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 16 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in e03edb73665d54dafca89ce853d832959b6452f2

**Review by tanwarsh — APPROVED** [review]
LGTM!

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 15 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in https://github.com/securefederatedai/openfl/commit/e03edb73665d54dafca89ce853d832959b6452f2

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 17 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: resolved in https://github.com/securefederatedai/openfl/commit/e03edb73665d54dafca89ce853d832959b6452f2

**Review by kagrawa2 — APPROVED** [review]
LGTM. Thanks @rajithkrishnegowda

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb [hunk 18]
```diff
@@ -0,0 +1,573 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "9a61e30e-83d1-422d-aa0a-ebd4376197ec",
+   "metadata": {},
+   "source": [
+    "# Scikit-learn Linear Regression Tutorial using Workflow Interface with Ridge Regularization\n",
+    "\n",
+    "\n",
+    "This tutorial demonstrates how to train a linear regression model using scikit-learn with Ridge regularization on a dataset, leveraging the new FedAI Workflow Interface. The Workflow Interface provides a novel way to compose federated learning experiments with OpenFL, enabling researchers to handle non-IID data and perform federated averaging. Through this tutorial, you will learn how to set up the federated learning environment, define the flow, and execute the training process across multiple collaborators."
```
Comment (teoparvanov) [inline openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: FedAI -> OpenFL

**Review by teoparvanov — APPROVED** [review]
This looks great, thank you @rajithkrishnegowda ! I just have a single comment on the overview cell, but otherwise this PR is ready to be merged.

File: openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb (hunk 18 above)
Comment (rajithkrishnegowda) [reply openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb]: updated in 3f9f447b6e9aa0d1023fe5ee5864bc0ac3a36f92

**Pull request merged**